components/places/src/storage/history.rs

```diff
@@ -446,7 +546,31 @@ pub mod history_sync {
             }
             None => new_page_info(db, &url, Some(incoming_guid.clone()))?,
         };
+
+        // Skip visits that are in tombstones, or that happen at the same time
+        // as visit that's already present. The 2nd lets us avoid inserting
+        // visits that we sent up to the server in the first place.
+        //
+        // It does cause us to ignore visits that legitimately happen
+        // at the same time, but that's probably fine and not worth
+        // worrying about.
+        let visits_to_skip: HashSet<Timestamp> = db.query_rows_into_cached(
+            "SELECT t.visit_date AS visit_date
+             FROM moz_historyvisit_tombstones t
+             WHERE place_id = :place_id
+             UNION ALL
+             SELECT v.visit_date AS visit_date
+             FROM moz_historyvisits v
+             WHERE place_id = :place_id",
```

Comment:
But this will return a lot of rows for a frequently-visited URL. How about adding a `visit_date IN` condition here that only returns timestamps for the visits we're adding?